src/main/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/transport/BatchInfoProvider.kt

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2024 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.common.schema.message.impl.rabbitmq.transport
+
+import kotlin.properties.Delegates
+
+/**
+ * This class holds information about book and session group from the batch root.
+ * It is used to pass this information down to [MessageId].
+ */
+internal sealed interface BatchInfoProvider {
+    val book: String
+    val sessionGroup: String
+
+    class Mutable : BatchInfoProvider {
+        override var book: String by Delegates.notNull()
+        override var sessionGroup: String by Delegates.notNull()
+        override fun equals(other: Any?): Boolean {
+            return this === other
+        }
+
+        override fun hashCode(): Int {
+            return System.identityHashCode(this)
```

Comment:
What is `compaction problem`? The identity hashcode is used because there is not internal state applicable for hash code. And there is not need to use this object as a key in map or add to a set

Reply:
Sorry I mean comparison problem.
So we expect that this object will be completed at the end of message parsing, also the values will be set only once. It's mean that we can calculate hash code when _object initialized_ otherwise hashCode and equal methods can throw an exception. The exception will be in runtime and we haven't got compilation check but we can do that.

Reply:
I believe the worst thing we can do is to throw an exception in equal method. There is no point in having equal and hashCode in this class use book and session group values - this class is not supposed to be used as a marker for equality. This is a simple holder. I can remove those definitions for equal and hashCode methods at all. And the JVM automatically applies the same behavior (only the same instances are equal). This is why I declared them explicitly here - a person can see that this class should not be used as equality marker